chain/lf3/participation_lease.go

```diff
@@ -0,0 +1,145 @@
+	maxLeasableInstances uint64
+}
+
+func newParticipationLeaser(nodeId peer.ID, progress gpbft.Progress, maxLeasedInstances uint64) *leaser {
+	return &leaser{
+		leases:               make(map[uint64]api.F3ParticipationLease),
+		issuer:               nodeId,
+		progress:             progress,
+		maxLeasableInstances: maxLeasedInstances,
+	}
+}
+
+func (l *leaser) getOrRenewParticipationTicket(participant uint64, previous api.F3ParticipationTicket, instances uint64) (api.F3ParticipationTicket, error) {
+
+	if instances > l.maxLeasableInstances {
+		return nil, api.ErrF3ParticipationTooManyInstances
+	}
+
+	currentInstance, _, _ := l.progress()
+	if len(previous) != 0 {
+		// A previous ticket is present. To avoid overlapping lease across multiple
+		// instances for the same participant check its validity and only proceed to
+		// issue a new ticket if:
+		//   - either it is expired/invalid, or
+		//   - it is valid and was issued by this node.
+		//
+		// Otherwise, return ErrF3ParticipationIssuerMismatch to signal to the caller the need for retry.
+		switch _, err := l.validate(currentInstance, previous); {
+		case errors.Is(err, api.ErrF3ParticipationTicketInvalid):
+			// Invalid ticket means the miner must have got the ticket from a node with a potentially different version.
+			// Refuse to issue a new ticket in case there is some other node with active lease for the miner.
+			return nil, err
+		case errors.Is(err, api.ErrF3ParticipationTicketExpired):
+			// The current instance is beyond the validity term of the previous lease. It is
+			// safe to proceed to issuing a ticket from current instance onwards for the term
+			// asked for.
+		case errors.Is(err, api.ErrF3ParticipationIssuerMismatch):
+			// The previous ticket is still valid and is not issued by this node; return error.
+			return nil, err
+		case errors.Is(err, api.ErrF3ParticipationTooManyInstances):
+			// We don't care if the previous lease was for too many instances. What we care
+			// about is that the new ticket is within the max which was checked right at the
+			// top.
+		case err != nil:
+			log.Errorw("Unexpected error occurred while validating previous participation ticket", "participant", participant, "err", err)
+			return nil, err
+		default:
+			// The previous ticket was issued by this node and is still valid. It is safe to
+			// proceed with issuing a new ticket with overlapping validity.
+		}
+		log.Debugw("Renewing previously issued participation ticket with overlapping lease", "participant", participant, "startInstance", currentInstance, "validFor", instances)
+	}
+
+	return l.newParticipationTicket(participant, currentInstance, instances)
+}
+
+func (l *leaser) participate(ticket api.F3ParticipationTicket) (api.F3ParticipationLease, error) {
+	currentInstance, _, _ := l.progress()
+	lease, err := l.validate(currentInstance, ticket)
+	if err != nil {
+		return api.F3ParticipationLease{}, err
+	}
+	l.mutex.Lock()
+	l.leases[lease.MinerID] = lease
```

Comment:
I think we need to look at the current lease. E.g., in case wires get crossed and an old call to participate gets stuck somewhere.

Storing an _expired_ lease is fine, we'll clean it up later. But clobbering a _good_ lease with an older lease is probably not something we want to do.

Reply:
> in case wires get crossed and an old call to participate gets stuck somewhere
> clobbering a good lease with an older lease

If we reach this line it means the the lease is "valid":
* decodes
* has not expired
* conforms to max leased instances accepted by node (that is controlled by Miner, see earlier feedback)

What specific edge case you have in mind Steb?

Reply:
IIRC, we allow slightly overlapping leases, right? We could end up in a situation where we:

1. Try to renew a lease for T1-T5. That gets stuck.
2. Try to renew the lease, now from T2-T6. That goes through.
3. The lease for T1-T5 goes through, replacing the longer lease.

I'm not sure if this is possible in the current architecture, but that's the kind of thing I'm a bit worried about.

Reply:
I see, interesting.

Over TCP connections I I am not sure if that's possible, unless there is some built-in retry in json rpc or at infra level that would blindly retry requests on behalf of nodes? I would then wonder how the response delivery would work. 

In any case, better safe than sorry. Thanks for pointing that out Steven

Reply:
1. Miner makes the first request, it gets all the way to the point where we'd record the lease then pauses.
2. Connection fails (miner reboots, etc.).
3. Miner makes the second request.
4. Lotus serves the second request.
5. The first request continues.

Reply:
I see thanks, you are right.

The latest commit should fix this such that the leaser never accepts a ticket with start instance before the previous lease. Please let know if I have missed anything :pray: